src/node_sets.cpp

```diff
@@ -0,0 +1,317 @@
+#include <algorithm>  // std::transform
+#include <cmath>
+#include <fmt/format.h>
+#include <sstream>
+#include <type_traits>
+
+#include <json.hpp>
+
+#include <bbp/sonata/node_sets.h>
+
+namespace bbp {
+namespace sonata {
+
+namespace {
+using json = nlohmann::json;
+
+const size_t MAX_COMPOUND_RECURSION = 10;
+
+template <typename T>
+std::string toString(const std::string& key, const std::vector<T>& values) {
+    return fmt::format(R"("{}": [{}])", key, fmt::join(values, ", "));
+}
+
+template <>
+std::string toString(const std::string& key, const std::vector<std::string>& values) {
+    // strings need to be wrapped in quotes
+    return fmt::format(R"("{}": ["{}"])", key, fmt::join(values, "\", \""));
+}
+
+template <typename T>
+class NodeSetBasicRule: public NodeSetRule
+{
+  public:
+    NodeSetBasicRule(std::string attribute, std::vector<T>&& values)
+        : attribute_(std::move(attribute))
+        , values_(values) {}
+
+    Selection materialize(const NodeSets& /* unused */, const NodePopulation& np) const final {
+        Selection ret{{}};
+        for (const auto& v : values_) {
+            ret = ret | np.matchAttributeValues(attribute_, v);
+        }
+        return ret;
+    }
+
+    std::string toJSON() const final {
+        return toString(attribute_, values_);
+    }
+
+  private:
+    std::string attribute_;
+    std::vector<T> values_;
+};
+
+class NodeSetBasicPopulation: public NodeSetRule
+{
+  public:
+    explicit NodeSetBasicPopulation(std::string name)
+        : name_(std::move(name)) {}
+
+    Selection materialize(const NodeSets& /* unused */, const NodePopulation& np) const final {
+        if (name_ == np.name()) {
+            return np.selectAll();
+        }
+
+        return Selection{{}};
+    }
+
+    std::string toJSON() const final {
+        const std::vector<std::string> values = {name_};
+        return toString("population", values);
+    }
+
+  private:
+    std::string name_;
+};
+
+class NodeSetBasicNodeIds: public NodeSetRule
+{
+  public:
+    explicit NodeSetBasicNodeIds(Selection::Values&& values)
+        : values_(values) {}
+
+    Selection materialize(const NodeSets& /* unused */,
+                          const NodePopulation& /* unused */) const final {
+        return Selection::fromValues(values_.begin(), values_.end());
```

Comment:
To me inside the nodesets all "[]" == "or". And I think it makes sense from a user point of view.
So to me,  for consistency, the node_id values should act like a "or" also and so you should only return the Selection of nodes actually inside the node population.

Reply:
Yup, makes sense.